In [28]:
import pandas as pd 
import numpy as np

### Reading the data

In [29]:
twb_dataset = pd.read_excel('../data/Copy of twb-chatbot-data (1).xlsx')
transperfect_dataset = pd.read_excel('../data/Covid-Utterances-en_US.xlsx')

In [30]:
twb_dataset.head()

ask_for_stats  \
0  What are the current COVID-19 statistics for DRC?   
1                                Statistics for DRC?   
2  What's the total number of people tested posit...   
3  How many people have tested positive for coron...   
4  If I am away from someone who is infected, I w...   

                                 covid_call_a_doctor  \
0                            Where can I get tested?   
1               What should I do if I have symptoms?   
2                       I have a dry cough and fever   
3                            Should I call a doctor?   
4  Are there people more at risk of being harmed ...   

                                     covid_disinfect  \
0               Can disinfectant hurt me in any way?   
1                     Do I have to use disinfectant?   
2               Is drinking disinfectant bad for me?   
3                       Should I drink disinfectant?   
4  Are there any differences between the coronavi...   

                              covid_explain_pandemic  \
0                                 What's a pandemic?   
1  What's the difference between a pandemic and a...   
2  What's the difference between a epidemic and a...   
3              Why is coronavirus called a pandemic?   
4  What do I do against feeling stress during the...   

                                     covid_gloves  \
0  Should I wear gloves to protect against COVID?   
1                       Does wearing gloves help?   
2                         Will gloves protect me?   
3                         Are gloves a good idea?   
4                                      Am I sick?   

                                      covid_hospital  \
0      Should I go to the hospital if I have corona?   
1       Do I need to visit a doctor if I have covid?   
2  Should I just stay at home if i have the disease?   
3  Is it okay to not go to the hospital if I have...   
4     Is there anything that cures the corona virus?   

                         covid_how_spread  \
0          How do people catch the virus?   
1        How do people catch the disease?   
2            How can I catch the disease?   
3            How does the disease spread?   
4  What's the minimum time to wash hands?   

                         covid_infection_sources  \
0  I wanted to know the origin not the treatment   
1          What is the origin of the coronavirus   
2                 What country is COVID-19 from?   
3                   Where does Corona come from?   
4          Is COVID-19 and coronavirus the same?   

                                  covid_masks_how_to  \
0                          How should I wear a mask?   
1  Was your mask made here at home or from other ...   
2              Do you have to wear a mask every day?   
3                        How should you wear a mask?   
4  Can I drink hot tea on the floor and not infec...   

                           covid_masks_effectiveness  ...  \
0                   What kind of mask should I wear?  ...   
1        Anyone who wears a mask can't get a corona?  ...   
2                           When does the mask help?  ...   
3  Does the mask send missiles to catch the coron...  ...   
4                                                NaN  ...   

       covid_protection_wash_hands  \
0  Why should you wash your hands?   
1             Why wash your hands?   
2    How should I wash my hands?\n   
3           How to wash your hands   
4                              NaN   

                      covid_protection_keep_distance  \
0  Do I have to be away from people to avoid corona?   
1                   How far should I be from people?   
2                      Should I be away from anyone?   
3                                                NaN   
4                                                NaN   

                             covid_risk_most_at_risk  \
0                Why do whites die more than blacks?   
1            Why don't crazy people get coronavirus?   
2  Is the cor

In [31]:
transperfect_dataset.head()

1. covid_intro   \
0  Are corona virus and COVID-19 different?   
1     are coronavirus and covid 19 the same   
2                  are coronaviruses common   
3                are coronaviruses seasonal   
4                       can covid-19 mutate   

                                     2. covid_origin  \
0  Are Bats really the original cause of the viru...   
1               Can you explain how Covid19 started?   
2             Can you explain how the virus started?   
3  Can you explain how this virus came into exist...   
4       Can you explain to me how the virus started?   

                                    3. symptoms_info  \
0                                      Am I infected   
1                   are coronavirus symptoms gradual   
2          are covid-19 antibody tests available yet   
3               are covid-19 stimulus checks taxable   
4  Are COVID-19 symptoms similar to that of the flu?   

                                   4. symtptoms_have  \
0                  Can the Corona virus cause fever?   
1            Can the fever I have, cause body aches?   
2  Can you tell me if having difficulties to brea...   
3  Can you tell me if my coughing and sneezing ar...   
4  Can you tell me if my difficulty to breathe is...   

                                5. emergency_contact  \
0  Are there any emergency numbers for the corona...   
1  Are there any means of contact for Covid19 pat...   
2  Are there any other ways besides email or phon...   
3                   Are there doctor numbers listed?   
4  Are these the only numbers available in case o...   

                                      6. precautions  \
0  "Social distancing"  means maintaining a dista...   
1                 Are masks effective against covid?   
2  Are masks important during the Covid19 virus o...   
3  Are masks the only option for preventing Covid19?   
4             Are there any way to stop corona virus   

                                  7. current_numbers  \
0  a corona virus mutates 100 x faster than influ...   
1            a map of the spread of the corona virus   
2                 a star singapore corona virus test   
3                          Amount of people infected   
4  Are all the countries in the world infected wi...   

                                       8. medication  \
0  Are fruits, vegetables and herbs useful in fig...   
1  Are there any existing medications I can take ...   
2  Are there any medicinal teas to help against t...   
3  Are there any Natural remedies that are effect...   
4  Are there medicines that make the virus more a...   

                                     9. myths_hoaxes  \
0               Are Africans immune against covid-19   
1  Are children under a certain age immune to the...   
2            are coronavirus victims getting younger   
3               Are young people immune to COVID-19?   
4  Can a mosquito spread the Corona virus with it...   

                                   10. how_spreading  
0  All outside contacts and physical proximity as...  
1  Are people around my dog at risk for corona vi...  
2         Ask around how covid-19 can be transmited.  
3                       Can animals spread covid-19?  
4             Can animals spread the covid-19 virus?

In [32]:
print(transperfect_dataset.shape, twb_dataset.shape)

(547, 10) (35, 25)


Here we can conclude that our dataset comes as a tabular form the first one has , the twb dataset has 547 row and 10 colums and the twbdataset has 35 colums.

let convert the dataset in the format readable to a machine learning model , the sentence will be be put in one colum and the second will be the intent ..

In [33]:
twb_dataset = pd.melt(twb_dataset, var_name='Intent', value_name='Question')

In [34]:
twb_dataset = twb_dataset[twb_dataset.Question.notna()]

In [35]:
twb_dataset.head()

Intent                                           Question
0  ask_for_stats  What are the current COVID-19 statistics for DRC?
1  ask_for_stats                                Statistics for DRC?
2  ask_for_stats  What's the total number of people tested posit...
3  ask_for_stats  How many people have tested positive for coron...
4  ask_for_stats  If I am away from someone who is infected, I w...

After doing some manipulation we have the dataset in the long format , with 2 columns one is the intent and the other one is the questions , and This is in the type \X, \Y format where the X is the question and the Y is the intent , this format is more suistable for any supervised machine learning technics.



We can do the same for the transperfect dataset to prefor the same operation on the twb dataset ...

In [36]:
transperfect_dataset = pd.melt(transperfect_dataset, var_name='Intent', value_name='Question')

In [37]:
transperfect_dataset = transperfect_dataset[transperfect_dataset.Question.notna()]

In [38]:
transperfect_dataset.shape

(2498, 2)

In [39]:
transperfect_dataset.head()

Intent                                  Question
0  1. covid_intro   Are corona virus and COVID-19 different?
1  1. covid_intro      are coronavirus and covid 19 the same
2  1. covid_intro                   are coronaviruses common
3  1. covid_intro                 are coronaviruses seasonal
4  1. covid_intro                        can covid-19 mutate

At this stage we have 2 differents datasets with labels ready for a machine learning task, we can start doing our NLP preprocessing , 

But before moving futher, let see if there are any similar column between the dataset ...

In [40]:
transperfect_dataset.Intent.value_counts()

7. current_numbers      547
3. symptoms_info        323
2. covid_origin         247
1. covid_intro          246
10. how_spreading       241
6. precautions          212
8. medication           194
5. emergency_contact    183
4. symtptoms_have       175
9. myths_hoaxes         130
Name: Intent, dtype: int64

In [41]:
twb_dataset.Intent.value_counts()

covid_treatments                  33
covid_symptoms                    25
ask_for_stats                     25
covid_protection                  21
covid_gloves                      19
covid_masks_effectiveness         17
covid_infection_sources           14
covid_masks_how_to                13
covid_what_is_corona_covid-19     10
covid_call_a_doctor               10
covid_myth_hot_tea                10
covid_risk_most_at_risk            9
covid_protection_wash_hands        9
covid_how_spread                   9
covid_hospital                     8
covid_wash_hands_how               7
covid_myth_antibiotics             7
covid_myth_UV                      7
covid_myth_only_old                7
covid_myth_mosquitos               7
covid_stress                       6
covid_SARS                         6
covid_explain_pandemic             6
covid_disinfect                    6
covid_protection_keep_distance     5
Name: Intent, dtype: int64

### Text Preprocessing 

In the following section we will perform text preprocessing for our both dataset and at the end of this section we will have our cleaned dataset were we can perform any datascience task.

We will use Spacy and NlTk package , here are the steps we will perform in this stage...

- Contraction removal , we will remove contraction words and replace what's == >  what is , it's == > it is , don't ==> do not , etc
-  The first step will be tokenization: this will consist of splitting each sentence into token, or word
-  Stemming or lemmatization will consist of replacing each word or token with their root word or word they derived.
-  Stop word removal
-  Spell checking  : here we will check words that are misspelled and correct them with the correction.
- We can also add name entity recognition techniques to detect entities such as person names and countries.

In [42]:
#! python -m spacy download en_core_web_sm

In [43]:
import spacy 
import nltk
import re
nlp = spacy.load('en_core_web_sm', parse = True, tag=True, entity=True)
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /Users/es.py/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

### Removing contractions 

From [this](https://towardsdatascience.com/a-practitioners-guide-to-natural-language-processing-part-i-processing-understanding-text-9f4abfd13e72) blog we can read :
Contractions are shortened version of words or syllables. They often exist in either written or spoken forms in the English language. These shortened versions or contractions of words are created by removing specific letters and sounds. In case of English contractions, they are often created by removing one of the vowels from the word. Examples would be, do not to don’t and I would to I’d. Converting each contraction to its expanded, original form helps with text standardization.


Thans to this repo, becasuse from [this](https://github.com/dipanjanS/practical-machine-learning-with-python/blob/master/bonus%20content/nlp%20proven%20approach/contractions.py) we can find list of all contraction s

In [44]:
CONTRACTION_MAP = {
"ain't": "is not",
"aren't": "are not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he would",
"he'd've": "he would have",
"he'll": "he will",
"he'll've": "he he will have",
"he's": "he is",
"how'd": "how did",
"how'd'y": "how do you",
"how'll": "how will",
"how's": "how is",
"I'd": "I would",
"I'd've": "I would have",
"I'll": "I will",
"I'll've": "I will have",
"I'm": "I am",
"I've": "I have",
"i'd": "i would",
"i'd've": "i would have",
"i'll": "i will",
"i'll've": "i will have",
"i'm": "i am",
"i've": "i have",
"isn't": "is not",
"it'd": "it would",
"it'd've": "it would have",
"it'll": "it will",
"it'll've": "it will have",
"it's": "it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"mightn't've": "might not have",
"must've": "must have",
"mustn't": "must not",
"mustn't've": "must not have",
"needn't": "need not",
"needn't've": "need not have",
"o'clock": "of the clock",
"oughtn't": "ought not",
"oughtn't've": "ought not have",
"shan't": "shall not",
"sha'n't": "shall not",
"shan't've": "shall not have",
"she'd": "she would",
"she'd've": "she would have",
"she'll": "she will",
"she'll've": "she will have",
"she's": "she is",
"should've": "should have",
"shouldn't": "should not",
"shouldn't've": "should not have",
"so've": "so have",
"so's": "so as",
"that'd": "that would",
"that'd've": "that would have",
"that's": "that is",
"there'd": "there would",
"there'd've": "there would have",
"there's": "there is",
"they'd": "they would",
"they'd've": "they would have",
"they'll": "they will",
"they'll've": "they will have",
"they're": "they are",
"they've": "they have",
"to've": "to have",
"wasn't": "was not",
"we'd": "we would",
"we'd've": "we would have",
"we'll": "we will",
"we'll've": "we will have",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what will",
"what'll've": "what will have",
"what're": "what are",
"what's": "what is",
"what've": "what have",
"when's": "when is",
"when've": "when have",
"where'd": "where did",
"where's": "where is",
"where've": "where have",
"who'll": "who will",
"who'll've": "who will have",
"who's": "who is",
"who've": "who have",
"why's": "why is",
"why've": "why have",
"will've": "will have",
"won't": "will not",
"won't've": "will not have",
"would've": "would have",
"wouldn't": "would not",
"wouldn't've": "would not have",
"y'all": "you all",
"y'all'd": "you all would",
"y'all'd've": "you all would have",
"y'all're": "you all are",
"y'all've": "you all have",
"you'd": "you would",
"you'd've": "you would have",
"you'll": "you will",
"you'll've": "you will have",
"you're": "you are",
"you've": "you have"
}

In [73]:
def check_mispelled_word(text):
    text = text.lower()
    text = text.replace('corona virus', 'coronavirus')
    text = text.replace('covid19', 'covid-19')
    text = text.replace('covid 19', 'covid-19')
    text = re.sub(r'\b{}\b'.format("corona"), "coronavirus", text)
    return text

In [74]:
def expand_contractions(text, contraction_mapping=CONTRACTION_MAP):
    
    contractions_pattern = re.compile('({})'.format('|'.join(contraction_mapping.keys())), 
                                      flags=re.IGNORECASE|re.DOTALL)
    def expand_match(contraction):
        match = contraction.group(0)
        first_char = match[0]
        expanded_contraction = contraction_mapping.get(match)\
                                if contraction_mapping.get(match)\
                                else contraction_mapping.get(match.lower())                       
        expanded_contraction = first_char+expanded_contraction[1:]
        return expanded_contraction
        
    expanded_text = contractions_pattern.sub(expand_match, text)
    expanded_text = re.sub("'", "", expanded_text)
    return expanded_text

In [75]:
expand_contractions("Y'all can't expand contractions I'd think")

'You all cannot expand contractions I would think'

all the step in we defined in the text preprocessing phase can be performed using this function...

In [76]:
def lemmatize_text(text):
    text = nlp(text)
    text = ' '.join([word.lemma_ if word.lemma_ != '-PRON-' else word.text for word in text])
    return text

In [77]:
def remove_stopwords(text, is_lower_case=False):
    tokens = nltk.word_tokenize(text)
    tokens = [token.strip() for token in tokens]
    if is_lower_case:
        filtered_tokens = [token for token in tokens if token not in stopwords.words()]
    else:
        filtered_tokens = [token for token in tokens if token.lower() not in stopwords.words()]
    filtered_text = ' '.join(filtered_tokens)    
    return filtered_text

In [78]:
def preprocess_text(text, contraction_expansion=True,
                     accented_char_removal=True, text_lower_case=True, 
                     text_lemmatization=True, special_char_removal=True, 
                     stopword_removal=True, remove_digits=True, check_mispelled_words=True):
        # expand contractions 
    if check_mispelled_words:
        text = check_mispelled_word(text)
    if contraction_expansion:
        text = expand_contractions(text)
        # lowercase the text    
    if text_lower_case:
        text = text.lower()
        # remove extra newlines
    doc = re.sub(r'[\r|\n|\r\n]+', ' ',text)
        # lemmatize text
    if text_lemmatization:
        text = lemmatize_text(text)
    text = re.sub(' +', ' ', text)
        # remove stopwords
    if stopword_removal:
        text = remove_stopwords(text, is_lower_case=text_lower_case)
    return text

In [79]:
twb_dataset['questions_cleanned'] = twb_dataset.Question.apply(preprocess_text, stopword_removal=False)
transperfect_dataset['questions_cleanned'] = transperfect_dataset.Question.apply(preprocess_text, stopword_removal=False)

In [80]:
twb_dataset.head()

Intent                                           Question  \
0  ask_for_stats  What are the current COVID-19 statistics for DRC?   
1  ask_for_stats                                Statistics for DRC?   
2  ask_for_stats  What's the total number of people tested posit...   
3  ask_for_stats  How many people have tested positive for coron...   
4  ask_for_stats  If I am away from someone who is infected, I w...   

                                  questions_cleanned  
0   what be the current covid-19 statistic for drc ?  
1                                statistic for drc ?  
2  what be the total number of people test positi...  
3  how many people have test positive for coronav...  
4  if i be away from someone who be infect , i wi...

In [81]:
transperfect_dataset.head()

Intent                                  Question  \
0  1. covid_intro   Are corona virus and COVID-19 different?   
1  1. covid_intro      are coronavirus and covid 19 the same   
2  1. covid_intro                   are coronaviruses common   
3  1. covid_intro                 are coronaviruses seasonal   
4  1. covid_intro                        can covid-19 mutate   

                        questions_cleanned  
0  be coronavirus and covid-19 different ?  
1     be coronavirus and covid-19 the same  
2                   be coronaviruse common  
3                 be coronaviruse seasonal  
4                      can covid-19 mutate

In [86]:
transperfect_dataset = transperfect_dataset.drop_duplicates(subset='Question')

At this stage we have a cleanned dataset , we can already start thinking about trainning our machine learning models 